mods/classic-desktop-icons.wh.cpp

```diff
@@ -0,0 +1,167 @@
+
+// Evil C preprocessor use but this gets very repetitive very fast
+#define YcSimpleFunctionHook(LIBRARY, FUNCTION)            \
+FARPROC YcTokenConcat(p, FUNCTION) = GetProcAddress(        \
+    LIBRARY,                                                \
+    #FUNCTION                                               \
+);                                                          \
+                                                            \
+if (!YcTokenConcat(p, FUNCTION))                            \
+{                                                           \
+    Wh_Log(L"Failed to get address of " #FUNCTION);         \
+    return FALSE;                                           \
+}                                                           \
+                                                            \
+if (!Wh_SetFunctionHook(                                    \
+    (void *)YcTokenConcat(p, FUNCTION),                     \
+    (void *)YcTokenConcat(FUNCTION, _hook),                 \
+    (void **)&YcTokenConcat(FUNCTION, _orig)                \
+)) Wh_Log(L"Failed to hook function " #FUNCTION " at address %p", YcTokenConcat(p, FUNCTION))
+#pragma endregion // "YcSimpleFunctionHook"
+
+struct
+{
+    BOOL background;
+} settings;
+
+HWND hDesktop;
+WNDPROC OriginalWndProc;
+
+#define LABELBG settings.background     \
+        ? GetSysColor(COLOR_BACKGROUND) \
+        : CLR_NONE
+
+void UpdateDesktop(void)
+{
+    if (hDesktop != NULL)
+    {
+        // Untheme the desktop
+        SetWindowTheme(hDesktop, NULL, NULL);
+        SendMessageW(hDesktop, WM_THEMECHANGED, 0, 0);
+
+        // Apply the desktop background color 
+        DWORD dwBgColor = LABELBG;
+
+        SendMessageW(hDesktop, LVM_SETTEXTBKCOLOR, NULL, (LPARAM)dwBgColor);
+    }
+}
+
+#define LVM_GETACCVERSION 0x10C1
+
+LRESULT CALLBACK NewWndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    if (uMsg == WM_SETREDRAW || uMsg == LVM_GETITEMCOUNT)
+    {
+        UpdateDesktop();
+    }
+
+    return OriginalWndProc(hWnd, uMsg, wParam, lParam);
+}
+
+HWND FindDesktopWindow(void)
+{
+    HWND baseWindow = FindWindowW(L"Progman", L"Program Manager");
+    if (baseWindow)
+    {
+        HWND defView = FindWindowExW(baseWindow, 0, L"SHELLDLL_DefView", NULL);
+        if (defView)
+        {
+            HWND desktop = FindWindowExW(defView, 0, L"SysListView32", NULL);
+
+            if (desktop)
+            {
+                return desktop;
+            }
+        }
+    }
+    return NULL;
+}
+
+void LoadSettings(void)
+{
+    settings.background = Wh_GetIntSetting(L"background");
+}
+
+BOOL Wh_ModInit(void) 
+{
+    Wh_Log(L"Initializing Classic Desktop Icons");
+    LoadSettings();
+
+    // Initially update the desktop.
+    hDesktop = FindDesktopWindow();
+    UpdateDesktop();
+
+    OriginalWndProc = (WNDPROC)GetWindowLongPtr(hDesktop, GWLP_WNDPROC);
+    SetWindowLongPtr(hDesktop, GWLP_WNDPROC, (LONG_PTR)NewWndProc);
```

Comment:
Using `SetWindowLongPtr` with `GWLP_WNDPROC` will make it incompatible with other mods that subclass the same window. You can include `<windhawk_utils.h>` and use:

* `WindhawkUtils::SetWindowSubclassFromAnyThread` for subclassing
* `WindhawkUtils::RemoveWindowSubclassFromAnyThread` for unsubclassing (in `Wh_ModUninit`)
* Change:\
`LRESULT CALLBACK NewWndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)`\
to:\
`LRESULT CALLBACK NewWndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam, UINT_PTR uIdSubclass, DWORD_PTR dwRefData)`
* Change:\
`return OriginalWndProc(hWnd, uMsg, wParam, lParam);`\
to:\
`return DefSubclassProc(hWnd, uMsg, wParam, lParam);`
* Remove `OriginalWndProc`

Reply:
Should be fixed now.